wit/streams.wit

```diff
@@ -209,18 +224,30 @@ interface streams {
     /// of the input stream is seen and all the data has been written to
     /// the output stream.
     ///
-    /// This function returns the number of bytes transferred.
+    /// This function returns the number of bytes transferred, and the status of
+    /// the output stream.
     forward: func(
         this: output-stream,
         /// The stream to read from
         src: input-stream
-    ) -> result<u64, stream-error>
+    ) -> result<tuple<u64, stream-status>>
 
     /// Create a `pollable` which will resolve once either the specified stream
-    /// is ready to accept bytes or the other end of the stream has been closed.
+    /// is ready to accept bytes or the `stream-state` has become closed.
+    ///
+    /// Once the stream-state is closed, this pollable is always ready
+    /// immediately.
+    ///
+    /// The created `pollable` is a child resource of the `output-stream`.
+    /// Implementations may trap if the `output-stream` is dropped before
+    /// all derived `pollable`s created with this function are dropped.
     subscribe-to-output-stream: func(this: output-stream) -> pollable
 
     /// Dispose of the specified `output-stream`, after which it may no longer
     /// be used.
+    /// Implementations may trap if this `output-stream` is dropped while
+    /// child `pollable` resources are still alive.
+    /// After this `output-stream` is dropped, implementations may report any
+    /// corresponding `input-stream` has `stream-state.closed`.
```

Comment:
Is this constraint specific to streams and their pollables, or do you see this as a general pattern where there is a parent/child relationship between resources?

For example, do you foresee a similar constraint on the Drop method of file descriptors, stating it may trap if a child `directory-entry-stream`, `input-stream` or `output-stream` is still open?

Reply:
The general pattern where one resource might be the child of another resource is something we aim to make possible to describe in the component model type system, but we aren't ready to do so in the preview 2 time frame.

We didn't find that this child constraint was appropriate for the streams of files. The way I drew the line was, "is this resource useful after its parent has been dropped?". With pollables of streams, readiness is telling you something particular about the stream itself: it doesn't mean anything without a stream to try reading from/writing to. On the other hand, with files, a descriptor is a "full" view of all the capabilities possible, but if you just care about reading out the contents of a directory or file, its reasonable to just "narrow" your view to the particular stream you care about, and drop the descriptor because you don't care about any other operations.

Reply:
Ah nice. Hadn't thought about it that way.

I completely agree.